c10/mobile/CPUProfilingAllocator.cpp

```diff
@@ -49,6 +31,65 @@ struct MemEvent {
     time(t), allocation_id(id), size(s), type(e) {}
 };
 
+bool overlaps(const MemBlock& a, const MemBlock& b) {
+  // two blocks dont overlap if
+  // |---a--------|--------------b--------|
+  // strat_a     end_a <= start_b       end_b
+  return
+    !((a.end_offset <= b.start_offset) || (b.end_offset <= a.start_offset));
+}
+
+bool validate_allocation_plan(
+    const std::vector<MemEvent>& alloc_events,
+    const std::vector<uint64_t>& allocation_offsets) {
+  std::set<MemBlock> allocations;
+  for (const auto& event : alloc_events) {
+    auto alloc_id = event.allocation_id;
+    // Skip allocations not managed by AllocationPlan
+    if (allocation_offsets[alloc_id] == std::numeric_limits<uint64_t>::max()) {
+      continue;
+    }
+    auto start_offset = allocation_offsets[alloc_id];
+    auto end_offset = allocation_offsets[alloc_id] + event.size;
+    MemBlock mem_block(start_offset, end_offset);
+    if (event.type == EventType::Allocate) {
+      // 1. Form a block whose start offset is end_offset of the candidate block.
+      // 2. Find lower_bound. Thus find iterator to the first block whose start offset
+      //    is not less than, i.e. >=, of this block.
+      //    i.e. first_it.start_offset >= mem_block.end_offset
+      //    all blocks after this iterator dont overlap with mem_block
+      // 3. Iterate over all blocks from the beginning till iterator and check
+      //    for overlaps.
+      // Still O(n)
+      const MemBlock query_block(
+          end_offset, std::numeric_limits<uint64_t>::max());
+      auto it = allocations.lower_bound(query_block);
+      if (it != allocations.end()) {
+        while(it != allocations.begin()) {
+          if (overlaps(*it, mem_block)) {
+            return false;
+          }
+          --it;
+        }
+        if (overlaps(*it, mem_block)) {
+          return false;
+        }
+      }
+      allocations.emplace(mem_block);
+    } else if (event.type == EventType::Free) {
+      auto it = allocations.find(mem_block);
```

Comment:
nit: you also need to validate the end of the range here. Because of your comparison operator `find` only finds the element based on the start

Reply:
So based on the previous comment regarding the invariant that the set contains only non-overlapping intervals, check based on start offset should suffice. And given allocations should always precede free, if we have found overlapping interval we would have returned from this function beforehand.
Although I might have missed something again.

Reply:
I wasn't referring to overlaps here, I was referring to sanity checking that the end of the interval you have in the set is the same as the end of the interval you're trying to remove.